### PR TITLE
diskutils: Add partprobe after partition creation

### PR DESCRIPTION
**Pull request by christopherco**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

* diskutils: Add partprobe after partition creation

There can be a timing issue where partition creation finishes but the
devtmpfs files are not populated in time for partition initialization.
So to deal with this, we call partprobe here to query and flush the
partition table information, which should enforce that the devtmpfs
files are created when partprobe returns control.

* diskutils: invoke partprobe with flock

Added flock because "partprobe -s" apparently doesn't always block.
flock is part of the util-linux package and helps to synchronize access
with other cooperating processes. The important part is it will block
if the fd is busy, and then execute the command. Adding a 5 second timeout
to prevent us from possibly waiting forever.

Signed-off-by: Chris Co <chrco@microsoft.com>

###### Does this affect the toolchain?  

NO

###### Test Methodology

Local build. Tested in QEMU environment where timing issue was occurring.
Also tested on most recent 1.0-dev branch where VHDX image builds started exhibiting this issue.

### CODE REVIEW

File: toolkit/tools/imagegen/diskutils/diskutils.go [hunk 1]
```diff
@@ -398,7 +398,19 @@ func CreateSinglePartition(diskDevPath string, partitionNumber int, partitionTab
...
 	}
-
+	// Update kernel partition table information
+	//
+	// There can be a timing issue where partition creation finishes but the
+	// devtmpfs files are not populated in time for partition initialization.
+	// So to deal with this, we call partprobe here to query and flush the
+	// partition table information, which should enforce that the devtmpfs
+	// files are created when partprobe returns control.
+	stdout, stderr, err := shell.Execute("partprobe", "-s", diskDevPath)
```
Comment (MateuszMalisz) [inline toolkit/tools/imagegen/diskutils/diskutils.go:408]: Which package does partprobe come from? We probably should keep track of external programs that we use...

Comment (christopherco) [reply toolkit/tools/imagegen/diskutils/diskutils.go:408]: `partprobe` comes from the `parted` package. Any recommendations on how best to track these external dependencies?

BTW I also have #731 available, which introduces a new `installutils.GetRequiredPackagesForInstall()`. Idea being that the packages returned from this list are required for running image creation. Currently this is a Mariner-specific package list. But we could maybe extend this idea to incorporate external dependencies for all our tools?

**Review by rlmenge — APPROVED** [review]

File: toolkit/tools/imagegen/diskutils/diskutils.go (hunk 1 above)
Comment (MateuszMalisz) [reply toolkit/tools/imagegen/diskutils/diskutils.go:408]: Not really. Preferably we'd package our toolkit as an RPM package and write Requires there ;)

File: toolkit/tools/imagegen/diskutils/diskutils.go [hunk 2]
```diff
@@ -378,6 +378,7 @@ func CreateSinglePartition(diskDevPath string, partitionNumber int, partitionTab
 	const (
 		fillToEndOption = "100%"
 		mibFmt          = "%dMiB"
+		timeout         = "5"
```
Comment (jiria) [inline toolkit/tools/imagegen/diskutils/diskutils.go:381]: Suggest calling it timeoutSecs to denote the unit for readability.

**Review by jiria — APPROVED** [review]

**Review by PawelWMS — APPROVED** [review]

File: toolkit/tools/imagegen/diskutils/diskutils.go [hunk 3]
```diff
@@ -398,7 +398,19 @@ func CreateSinglePartition(diskDevPath string, partitionNumber int, partitionTab
...
+	// devtmpfs files are not populated in time for partition initialization.
+	// So to deal with this, we call partprobe here to query and flush the
+	// partition table information, which should enforce that the devtmpfs
+	// files are created when partprobe returns control.
+	stdout, stderr, err := shell.Execute("partprobe", "-s", diskDevPath)
+	if err != nil {
+		logger.Log.Warnf("Failed to execute partprobe: %v", stderr)
+		return "", err
+	}
+	logger.Log.Debugf(stdout)
```
[the remark this answers is not in the document]
Comment (christopherco) [reply toolkit/tools/imagegen/diskutils/diskutils.go:420]: `shell.Execute` already logs the exact invocation at DEBUG level. The debug output currently looks like:
```
DEBU[0004] Executing: [flock --timeout 5 /dev/loop5 partprobe -s /dev/loop5]
DEBU[0004] /dev/loop5: gpt partitions 1
```

Will add a bit more to the Debugf print for clarity.

**Review by jslobodzian — APPROVED** [review]

**Pull request merged**